structuralcodes/core/base.py

```diff
@@ -102,9 +104,50 @@ def get_ultimate_strain(self) -> t.Tuple[float, float]:
     def __marin__(self, **kwargs):
         """Function for getting the strain limits and coefficients
         for marin integration.
+
+        By default the law is discretized as a piecewise linear
+        function. Then marin coefficients are computed based on this
+        discretization.
         """
-        del kwargs
-        raise NotImplementedError
+
+        # Discretize the constitutive law in a "smart way"
+        def find_x_lim(x, y):
+            # Find the index where x changes from negative to non-negative
+            zero_crossing_index = np.argmax(x >= 0)
+
+            # Check if there are non-zero values for x > 0
+            if np.any(y[zero_crossing_index:] != 0):
+                # Find the last non-zero index for x > 0
+                non_zero_indices = np.nonzero(y[zero_crossing_index:])[0]
+                x_lim_index = zero_crossing_index + non_zero_indices[-1]
+                return x[x_lim_index]
+            # All values are zero for x > 0
+            return None
+
+        eps_max, eps_min = self.get_ultimate_strain()
+        if eps_max > 1:
+            eps_max = 1
+        eps = np.linspace(eps_min, eps_max, 100000)
+        sig = self.get_stress(eps)
+        sig[(sig > 0) & (sig < np.max(sig) * 1e-6)] = 0
```

Comment:
Please comment why L132 is necessary.

Reply:
We need this line for the logic of the "smart" linearization. 
It may happen that we will have a positive branch in the constitutive law. For exponential for instance we have an exponential decay so the stress is never going to zero, but tend to it.
With L132 I set to 0 very small numbers of stress in the positive branch. This is done because the method find_x_lim finds the index where stress is back to zero again in order to discretize only the meaningful part of the stress-strain diagram.

Reply:
Got it. Great if we could also add this check on the negative branch.